engine/src/main/java/io/zeebe/engine/nwe/task/ReceiveTaskProcessor.java

```diff
@@ -0,0 +1,99 @@
+  private final BpmnEventSubscriptionBehavior eventSubscriptionBehavior;
+
+  public ReceiveTaskProcessor(final BpmnBehaviors behaviors) {
+    eventSubscriptionBehavior = behaviors.eventSubscriptionBehavior();
+    incidentBehavior = behaviors.incidentBehavior();
+    stateBehavior = behaviors.stateBehavior();
+    stateTransitionBehavior = behaviors.stateTransitionBehavior();
+    variableMappingBehavior = behaviors.variableMappingBehavior();
+  }
+
+  @Override
+  public Class<ExecutableReceiveTask> getType() {
+    return ExecutableReceiveTask.class;
+  }
+
+  @Override
+  public void onActivating(final ExecutableReceiveTask element, final BpmnElementContext context) {
+    variableMappingBehavior
+        .applyInputMappings(context, element)
+        .flatMap(ok -> eventSubscriptionBehavior.subscribeToEvents(element, context))
+        .ifRightOrLeft(
+            ok -> stateTransitionBehavior.transitionToActivated(context),
+            failure -> incidentBehavior.createIncident(failure, context));
+  }
+
+  @Override
+  public void onActivated(final ExecutableReceiveTask element, final BpmnElementContext context) {
+    // nothing to do here
+  }
+
+  @Override
+  public void onCompleting(final ExecutableReceiveTask element, final BpmnElementContext context) {
+    variableMappingBehavior
+        .applyOutputMappings(context, element)
+        .ifRightOrLeft(
+            ok -> {
+              eventSubscriptionBehavior.unsubscribeFromEvents(context);
+              stateTransitionBehavior.transitionToCompleted(context);
+            },
+            failure -> incidentBehavior.createIncident(failure, context));
+  }
+
+  @Override
+  public void onCompleted(final ExecutableReceiveTask element, final BpmnElementContext context) {
+    stateTransitionBehavior.takeOutgoingSequenceFlows(element, context);
+    stateBehavior.consumeToken(context);
+    stateBehavior.removeElementInstance(context);
+  }
+
+  @Override
+  public void onTerminating(final ExecutableReceiveTask element, final BpmnElementContext context) {
+    eventSubscriptionBehavior.unsubscribeFromEvents(context);
+    stateTransitionBehavior.transitionToTerminated(context);
+  }
+
+  @Override
+  public void onTerminated(final ExecutableReceiveTask element, final BpmnElementContext context) {
+    eventSubscriptionBehavior.publishTriggeredBoundaryEvent(context);
+    incidentBehavior.resolveIncidents(context);
+    stateTransitionBehavior.onElementTerminated(element, context);
+    stateBehavior.consumeToken(context);
+  }
+
+  @Override
+  public void onEventOccurred(
+      final ExecutableReceiveTask element, final BpmnElementContext context) {
+    if (eventSubscriptionBehavior.hasBoundaryEventOccurred(element, context)) {
+      eventSubscriptionBehavior.triggerBoundaryEvent(element, context);
+    } else {
+      eventSubscriptionBehavior.triggerIntermediateEvent(context);
+      stateTransitionBehavior.transitionToCompleting(context);
```

Comment:
Looking at `triggerBoundaryEvent()` and `triggerIntermediateEvent()`, I have the feeling that `transitionToCompleting()` can be part of `triggerIntermediateEvent()`. In case of an interrupting boundary event, the transition is also taken inside the `eventSubscriptionBehavior`.

Maybe, we can also combine `triggerBoundaryEvent()` and `triggerIntermediateEvent()` in the `eventSubscriptionBehavior` to peek the event trigger only once.